lib/ansible/executor/task_queue_manager.py

```diff
@@ -185,8 +188,60 @@ def __init__(
         # plugins for inter-process locking.
         self._connection_lockfile = tempfile.TemporaryFile()
 
+        self._workers: list[WorkerProcess | None] = []
+
+        # signal handlers to propagate signals to workers
+        signal.signal(signal.SIGTERM, self._signal_handler)
+        signal.signal(signal.SIGINT, self._signal_handler)
+
     def _initialize_processes(self, num: int) -> None:
-        self._workers: list[WorkerProcess | None] = [None] * num
+        # mutable update to ensure the reference stays the same
+        self._workers[:] = [None] * num
+
+    def _signal_handler(self, signum, frame) -> None:
+        """
+        terminate all running process groups created as a result of calling
+        setsid from within a WorkerProcess.
+
+        Since the children become process leaders, signals will not
+        automatically propagate to them.
+        """
+        signal.signal(signum, signal.SIG_DFL)
+        signame = signal.Signals(signum).name
+
+        display.debug(f'Handling {signame} signal')
+
+        alive = []
+        for worker in self._workers:
+            if worker is None or not worker.is_alive():
+                continue
+            if worker.pid:
+                alive.append(worker)
+                display.debug(f'Propagating {signame} signal to WorkerProcess[{worker.pid}]')
+                try:
+                    # notify workers
+                    os.kill(worker.pid, signum)
```

Comment:
If we do decide to remove special handling from the `WorkerProcess`, this would need to become `os.killpg`.